### DISCUSSION TITLE
How to get features with stumpy?

### DISCUSSION THREAD

**Discussion by b-y-f** [TDAmeritrade/stumpy#740] [Help: Coding & Implementations]
I got a [question](https://datascience.stackexchange.com/questions/116348/feature-engineering-time-series-data-to-capture-a-specific-pattern-or-pay-more-w) about extractor from time-series and someone answered me to use MatrixProfile, then I come here, does anyone knows how can I get some feature that pays extra weight to the motif and how to find that motif if I know it is in the plot???

**Comment by seanlaw** [comment]
> does anyone knows how can I get some feature that pays extra weight to the motif and how to find that motif if I know it is in the plot?

@b-y-f Can you please provide more details? I'm not able to understand your point

**Reply by b-y-f** [reply]
Hi @seanlaw ,

We are experiencing low recall and accuracy when using the ensemble method to classify time series data. And I'm curious if we can solve it with stumpy; currently, we're using traditional methods to extract features for training time-series data; one column is 30m `time-series`, another column is `value`, and after extracting the feature, we got some mean, maximum, and lagged features.

We want to use MatrixProfile to extract some motifs from features to help us improve the classification model's performance. For example, in our case, we know there is a motif of charging a battery before training the model. We also received labels for each motif, so we want to use the previous knowledge of charging patterns to improve the model.
<img width="336" alt="image" src="https://user-images.githubusercontent.com/13849223/204441933-08d0b6c6-be0d-4c04-a37c-796bb156d40d.png">
This TS data from our dataset, we want capture those motifs and convert them to features for training the timeseries model.
![image](https://user-images.githubusercontent.com/13849223/204443057-ff95f11e-2bd4-4f7f-94f6-a1921084f4e6.png)


And want to know with **stumpy** there exist any API to help us do that?

**Reply by seanlaw** [reply]
The short answer is "maybe" as I don't fully know what you are trying to achieve. I recommend taking a look at our [Shapelet Discovery Tutorial](https://stumpy.readthedocs.io/en/latest/Tutorial_Shapelet_Discovery.html). If you are unfamiliar with matrix profiles, then I recommend taking a look at this [video intro](https://stumpy.readthedocs.io/en/latest/motivation.html).

To be absolutely clear, matrix profile outputs may be used for machine learning but that is/was not the original intended use case.

**Reply by b-y-f** [reply]
> To be absolutely clear, matrix profile outputs may be used for machine learning but that is/was not the original intended use case.

Thanks, looks like we can validate our prediction data with MatrixProfile to find motif(we don't have any label for predicted data). It seems a good tool to detect patterns visually in a single or few time-series.

I see some usage of Shaplet for classification, in that tutorial article, it uses a different value as a feature vector to fit the model.

**Comment by NimaSarajpoor** [comment]
@b-y-f 
If I understand correctly, you want to create a model that takes two inputs: (1) X = the time series `T` with length `n` and (2) `y= labels with type binary` of length `n` , and you want to predict the label for a new time stamp. 

Am I right? If yes, please continue. I have a few ideas that MIGHT help you..

(0) If you are completely new to matrix profile, you may want to learn its basics, Also: learn self-join vs AB-join 

(1) As @seanlaw said, you should take  a look at the Shapelet Discovery Tutorial. Then, I believe you need to break down your time series `T` into chunks, where the time stamps in each chunk have the same labels. So, you will have bunch of chunks and each chunk has one label now. You can do this on your train data. Then use shapelet to build classifier. 

If the chunks in (1) do not have equal length, then you may want to use:

* scaling technique: This is newly added to [docs](https://github.com/TDAmeritrade/stumpy/tree/main/docs). The tutorial is [Tutorial_Uniform_Scaling.ipynb](https://github.com/TDAmeritrade/stumpy/blob/main/docs/Tutorial_Uniform_Scaling.ipynb). Again, this is new, so I know a little bit about it. So, feel free to explore it.

* Write a code that uses elastic distance: DTW, K-Shape, etc that can calculate the distance between two sequence with different lengths. However, note that this can increase the computing time as it does not take advantage of the matrix profile algorithm which uses Euclidean distance.


(2) Another approach is to create tabular data, e.g. by using lags. I am not 100% sure but I think this video might explain that: https://www.youtube.com/watch?v=9QtL7m3YS9I

**Reply by b-y-f** [reply]
> If I understand correctly, you want to create a model that takes two inputs: (1) X = the time series T with length n and (2) y= labels with type binary of length n , and you want to predict the label for a new time stamp.

Actually, we want to use the a value and it's derivative features to predict the label y with binary type.
We got some data that got the label and most of the data did not have that label.

> (2) Another approach is to create tabular data, e.g. by using lags. I am not 100% sure but I think this video might explain that: https://www.youtube.com/watch?v=9QtL7m3YS9I

Yes, we did exactly what this video said to the data, so we have 40+ features. I will give you a sample data after the feature engineered. ( f1,f2...are derived from original value)
```
ts  original_value     f1    f2   label
------------------------
1    2000   a     b    True
2    5000   c     d    True
3    4500  e     f    False
4    10  c     g    True
5    100  y     r    True
......

**Reply by b-y-f** [reply]
> I believe you need to break down your time series T into chunks, where the time stamps in each chunk have the same labels.

Do you mean with this kind of method? https://youtu.be/BYjOp2NoDdc?t=1770

**Reply by NimaSarajpoor** [reply]
Sorry for the late response as I have been busy lately. Before I jump into the technical stuff, please allow me to give you a suggestion. I think you should try to explain your question better/simpler. For instance, you said:

> Actually, we want to use the a value and it's derivative features to predict the label y with binary type.

However, if I understand correctly, the derived features are not given! YOU derive those features and use them. Right? If yes, then in my opinion, you should have just said: "we have this input that comes with label." I mean....there is no need to talk about derived features here.

Another example: when you were responding to @JaKasb, you said:

> each feature vector is the distance of each time series to all other meters?

what is "each time series"? Did you mean "each subsequence"? 

What "meter"? If it is not important, you should have avoided using it as it creates confusion for the reader. 

My only goal is to help you to better deliver your message to the reader. For instance, let's compare the two following versions of a question:

**version I)** I have a wind farm that generate 23.2 MW power at 3pm on 04-Nov-2022, and I also have another wind farm that generates 12.4 MW power at 3pm on 04-Nov-2022. Can someone help me calculate how much power these two wind farms generate together?

**version II)** If I have `X` and `Y`, how can I calculate their sum? 

As you can see the second version is easier to understand! Note that I used my domain knowledge to simplify the question in the second version! I KNOW that when two generators generate together, their outputs should be added. So, now, I can reduce the complexity of the question and make it simpler.

----

Okay, now let's get back to your problem...

> I will give you a sample data after the feature engineered. ( f1,f2...are derived from original value)

I assume you derive those features and they did not initially exist. Is that correct? If yes, then please remove them. Can you try to drag & drop a csv file here that contains your data?



> and most of the data did not have that label.

Well, this is a new information! The first thing that came to my mind was: "okay...if there is no label, how can one evaluate the prediction?" I think if you provide your data here, it can help us to understand what challenge you would like to tackle. 


> > I believe you need to break down your time series T into chunks, where the time stamps in each chunk have the same labels.
> 
> Do you mean with this kind of method? https://youtu.be/BYjOp2NoDdc?t=1770

I did not check the whole video. But, it seems their data (electricity demand) does not have label, which is different than your problem. Is that correct? What I was talking about is something different. I explain it below:

`time series = [a, b, c, d, e, f, g, h, i, ...]`
`label = [0, 0, 1, 1, 1,, 0, 0, 1, 1, ...]`

Then: 
chunks (a.k.a slices) with `label==1` are: `[c, d, e]`, `[h, i]`
chunks (a.k.a slices) with `label==0` are: `[a, b]`, `[f, g]`

Then, you MIGHT be able to use these chunks for the training stage in the shapelet discovery. Again, this is just an opinion. I have not explored this path before!

**Reply by b-y-f** [reply]
> As you can see the second version is easier to understand! Note that I used my domain knowledge to simplify the question in the second version! 

Thanks for the suggestion, will improve my question to make it more concise! :)


> Can you try to drag & drop a csv file here that contains your data?


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/yifan/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/yifan/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



delivered_wh | received_wh | hour_minute | year | day | month | lagged_1_delivered_wh | lagged_2_delivered_wh | lagged_3_delivered_wh | change_rate_delivered_wh | week_day_numeric | week_day_nominal | mean_delivered_wh_3hr | mean_received_wh_3hr | max_delivered_wh_3hr | max_received_wh_3hr | min_delivered_wh_3hr | min_received_wh_3hr | 1h1wMINUSwh | 1h1dMINUSwh | 2h1wMINUSwh | 2h1dMINUSwh | 1h2wMINUSwh | 1h2dMINUSwh | 2h2wMINUSwh | 2h2dMINUSwh | 1h3wMINUSwh | 1h3dMINUSwh | 2h3wMINUSwh | 2h3dMINUSwh | 1h4wMINUSwh | 1h4dMINUSwh | 2h4wMINUSwh | 2h4dMINUSwh | mean1h1_2_3_4wMINUSwh | median1h1_2_3_4wMINUSwh | mean1h1_2_3_4dayMINUSwh | median1h1_2_3_4day | mean2h1_2_3_4wMINUSwh | median2h1_2_3_4wMINUSwh | mean2h1_2_3_4dayMINUSwh | median2h1_2_3_4day | EV_charging_prediction | EV_charging_False_confidence
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
2204 | 26 | 13 | 2021 | 1 | 10 | 0 | 0 | 0 | 0 | 5 | Friday | 2204 | 26 | 2204 | 26 | 2204 | 26 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.68014105
3262 | 0 | 13.5 | 2021 | 1 | 10 | 2204 | 0 | 0 | 0.3243409 | 5 | Friday | 2733 | 13 | 3262 | 26 | 2204 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.5909775
3992 | 0 | 14 | 2021 | 1 | 10 | 3262 | 2204 | 0 | 0.18286573 | 5 | Friday | 3152.66667 | 8.66666667 | 3992 | 26 | 2204 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.5268584
2470 | 12 | 14.5 | 2021 | 1 | 10 | 3992 | 3262 | 2204 | -0.3812625 | 5 | Friday | 2982 | 9.5 | 3992 | 26 | 2204 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.76679584
908 | 6 | 15 | 2021 | 1 | 10 | 2470 | 3992 | 3262 | -0.6323887 | 5 | Friday | 2567.2 | 8.8 | 3992 | 26 | 908 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.91198459
388 | 2 | 15.5 | 2021 | 1 | 10 | 908 | 2470 | 3992 | -0.5726872 | 5 | Friday | 2204 | 7.66666667 | 3992 | 26 | 388 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.91198459
254 | 0 | 16 | 2021 | 1 | 10 | 388 | 908 | 2470 | -0.3453608 | 5 | Friday | 1879 | 3.33333333 | 3992 | 12 | 254 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
990 | 0 | 16.5 | 2021 | 1 | 10 | 254 | 388 | 908 | 0.74343434 | 5 | Friday | 1500.33333 | 3.33333333 | 3992 | 12 | 254 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
1044 | 0 | 17 | 2021 | 1 | 10 | 990 | 254 | 388 | 0.05172414 | 5 | Friday | 1009 | 3.33333333 | 2470 | 12 | 254 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
26 | 12 | 17.5 | 2021 | 1 | 10 | 1044 | 990 | 254 | -0.9750958 | 5 | Friday | 601.666667 | 3.33333333 | 1044 | 12 | 26 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
16 | 10 | 18 | 2021 | 1 | 10 | 26 | 1044 | 990 | -0.3846154 | 5 | Friday | 453 | 4 | 1044 | 12 | 16 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
10 | 12 | 18.5 | 2021 | 1 | 10 | 16 | 26 | 1044 | -0.375 | 5 | Friday | 390 | 5.66666667 | 1044 | 12 | 10 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
226 | 32 | 19 | 2021 | 1 | 10 | 10 | 16 | 26 | 0.95575221 | 5 | Friday | 385.333333 | 11 | 1044 | 32 | 10 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
86 | 34 | 19.5 | 2021 | 1 | 10 | 226 | 10 | 16 | -0.619469 | 5 | Friday | 234.666667 | 16.6666667 | 1044 | 34 | 10 | 0 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
106 | 12 | 20 | 2021 | 1 | 10 | 86 | 226 | 10 | 0.18867925 | 5 | Friday | 78.3333333 | 18.6666667 | 226 | 34 | 10 | 10 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
8 | 18 | 20.5 | 2021 | 1 | 10 | 106 | 86 | 226 | -0.9245283 | 5 | Friday | 75.3333333 | 19.6666667 | 226 | 34 | 8 | 10 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207
70 | 14 | 21 | 2021 | 1 | 10 | 8 | 106 | 86 | 0.88571429 | 5 | Friday | 84.3333333 | 20.3333333 | 226 | 34 | 8 | 12 | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | NaN | 0 | 0.98480207



</body>

</html>

**Comment by JaKasb** [comment]
1. Identify Recurring Time Series Patterns
It seems you already know your desired shapes.
Each shape is a query in a pattern similarity search.
As an alternative you can use an Shapelets / Contrast Profiles, if you have a labelled training set.
Or you could try regular motif search to find repeating patterns and use the found motifs as queries.

2. Compute "Feature Vectors"
Your desired feature vector is a time series of distances or correlations.
You can use MASS or DTW to compute the sliding window pattern similarity.
Repeat for each search query pattern.

3. Interpretation
The distances/correlation time series are the input for a binary decision algorithm.
If you have labeled data, then you can use any binary classification algorithm.

**Reply by b-y-f** [reply]
> You can use MASS or DTW to compute the sliding window pattern similarity.

Interesting! Do you mean we can extract feature vectors from sliding window DTW, e.g. each feature vector is the distance of each time series to all other meters?

**Reply by JaKasb** [reply]
It is not clear to me what you mean by "each time series" and "meters".

feature_vector_i = MASS(known_query_pattern_i, long_timeseries_corpus)
repeat for each known_query_pattern.

In the end, you *need* to experiment with functions and visualize the feature_vectors.
Most likely you will need an additional postprocessing function for the raw distance vectors, such as thresholding.

See
https://www.cs.unm.edu/~mueen/FastestSimilaritySearch.html
https://stumpy.readthedocs.io/en/latest/search.html?q=mass

**Reply by b-y-f** [reply]
> It is not clear to me what you mean by "each time series" and "meters".

We got many IoT devices each IoT's data is a "time series" dataset and we call each devices "meters".
